fggs/indices.py

```diff
@@ -0,0 +1,1293 @@
+        yield self
+
+    def forward(self, subst: Subst) -> Embedding:
+        """Look in subst for the end of the forwarding chain starting with self."""
+        return self
+
+    @abstractmethod
+    def clone(self, subst: Subst) -> Embedding:
+        """Apply the given substitution, creating new Embedding objects except
+           reuse existing EmbeddingVar objects."""
+        pass
+
+    @abstractmethod
+    def alpha(e, f: Embedding, rename: Rename) -> bool:
+        """Check whether the given renaming turns e into f.  Each EmbeddingVar
+           in e must be a key in the given renaming, even if it's also in f."""
+        pass
+
+    @abstractmethod
+    def freshen(self, rename: Rename) -> Embedding:
+        """Copy self while performing the given renaming, extending the
+           renaming whenever necessary."""
+        pass
+
+    @abstractmethod
+    def reassociate(self) -> Iterator[Embedding]:
+        """Remove EmbeddingVars of size 1, and merge nested ProductEmbeddings."""
+        pass
+
+    def unify(e: Embedding, f: Embedding, subst: Subst) -> bool:
+        """Unify the two embeddings by extending subst. Return success."""
+        e = e.forward(subst)
+        f = f.forward(subst)
+        if e is f: return True
+        if isinstance(e, ProductEmbedding) and isinstance(f, ProductEmbedding):
+            if len(e.factors) == len(f.factors):
+                return all(e1.unify(f1, subst) for e1, f1 in zip(e.factors, f.factors))
+            else:
+                warn(f"Attempt to unify {e.depict(debugging_letterer)} and {f.depict(debugging_letterer)} indicates index type mismatch")
+                return False
+        if isinstance(e, SumEmbedding) and isinstance(f, SumEmbedding):
+            if e.before == f.before and e.after == f.after:
+                return e.term.unify(f.term, subst)
+            else:
+                if __debug__:
+                    etn = e.term.numel()
+                    ftn = f.term.numel()
+                    if e.before + etn + e.after != f.before + ftn + f.after \
+                       or e.before + etn > f.before and f.before + ftn > e.before:
+                        warn(f"Attempt to unify {e.depict(debugging_letterer)} and {f.depict(debugging_letterer)} indicates index type mismatch")
+                return False
+        if __debug__:
+            if e.numel() != f.numel():
+                warn(f"Attempt to unify {e.depict(debugging_letterer)} and {f.depict(debugging_letterer)} indicates index type mismatch")
+        if isinstance(e, EmbeddingVar):
+            subst[e] = f
+            return True
+        if isinstance(f, EmbeddingVar):
+            subst[f] = e
+            return True
+        warn(f"Attempt to unify {e.depict(debugging_letterer)} and {f.depict(debugging_letterer)} indicates index type mismatch")
+        return False
+
+    def antiunify(e: Embedding, f: Embedding, antisubst: AntiSubst) -> Embedding:
```

Comment:
I'm not familiar with antiunification and am guessing a lot of other people also aren't